File: go/dsa/list/DoublyLinkedList/DoublyLinkedList.go
package doublylinkedlist

import "fmt"

// ListNode representa um nó na lista duplamente encadeada
type ListNode struct {
	Data     int       // Dado armazenado no nó
	Next     *ListNode // Ponteiro para o próximo nó na lista
	Previous *ListNode // Ponteiro para o nó anterior na lista
}

// DoublyLinkedList representa a estrutura da lista duplamente encadeada
type DoublyLinkedList struct {
	head   *ListNode // Ponteiro para o primeiro nó da lista
	tail   *ListNode // Ponteiro para o último nó da lista
	length int       // Contador do número de nós na lista
}

// NewDoublyLinkedList cria e retorna uma nova lista duplamente encadeada
func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{}
}

// IsEmpty verifica se a lista está vazia
func (dll *DoublyLinkedList) IsEmpty() bool {
	return dll.length == 0
}

// Length retorna o comprimento da lista
func (dll *DoublyLinkedList) Length() int {
	return dll.length
}

// DisplayForward exibe a lista do início ao fim
func (dll *DoublyLinkedList) DisplayForward() {
	if dll.head == nil {
		return
	}

	temp := dll.head
	for temp != nil {
		fmt.Printf("%d --> ", temp.Data)
		temp = temp.Next
	}
	fmt.Println("null")
}

// DisplayBackward exibe a lista do fim ao início
func (dll *DoublyLinkedList) DisplayBackward() {
	if dll.tail == nil {
		return
	}

	temp := dll.tail
	for temp != nil {
		fmt.Printf("%d --> ", temp.Data)
		temp = temp.Previous
	}
	fmt.Println("null")
}

// InsertFirst insere um nó no início da lista
func (dll *DoublyLinkedList) InsertFirst(value int) {
	newNode := &ListNode{Data: value}
	if dll.IsEmpty() {
		dll.tail = newNode
	} else {
		dll.head.Previous = newNode
	}
	newNode.Next = dll.head
	dll.head = newNode
	dll.length++
}

// InsertEnd insere um nó no fim da lista
func (dll *DoublyLinkedList) InsertEnd(value int) {
	newNode := &ListNode{Data: value}
	if dll.IsEmpty() {
		dll.head = newNode
	} else {
		dll.tail.Next = newNode
		newNode.Previous = dll.tail
	}
	dll.tail = newNode
	dll.length++
}

// DeleteFirst remove o primeiro nó da lista e o retorna
func (dll *DoublyLinkedList) DeleteFirst() *ListNode {
	if dll.IsEmpty() {
		panic("Doubly Linked List is already empty")
	}

	temp := dll.head
	if dll.head == dll.tail {
		dll.tail = nil
	} else {
		dll.head.Next.Previous = nil
	}
	dll.head = dll.head.Next
	temp.Next = nil
	dll.length--
	return temp
}

// DeleteLast remove o último nó da lista e o retorna
func (dll *DoublyLinkedList) DeleteLast() *ListNode {
	if dll.IsEmpty() {
		panic("Doubly Linked List is already empty")
	}

	temp := dll.tail
	if dll.head == dll.tail {
		dll.head = nil
	} else {
		dll.tail.Previous.Next = nil
	}
	dll.tail = dll.tail.Previous
	temp.Previous = nil
	dll.length--
	return temp
}

File: csharp/src/dsa/list/DoublyLinkedList.cs
namespace dsa.list
{
    public class DoublyLinkedList
    {
        // static void Main()
        // {
        //     DoublyLinkedList dll = new DoublyLinkedList();
        //     dll.insertEnd(1);
        //     dll.insertEnd(2);
        //     dll.insertEnd(3);

        //     dll.displayForward();

        //     dll.deleteLast();
        //     dll.deleteLast();

        //     dll.displayForward();
        // }
        
        // Ponteiro para o primeiro nó da lista
        private ListNode head;
        // Ponteiro para o último nó da lista
        private ListNode tail;
        // Contador do número de nós na lista
        private int _length;

        // Classe interna para representar um nó na lista
        public class ListNode
        {
            // Dado armazenado no nó
            public int data;
            // Ponteiro para o próximo nó na lista
            public ListNode next;
            // Ponteiro para o nó anterior na lista
            public ListNode previous;

            // Construtor do nó
            public ListNode(int data)
            {
                this.data = data;
            }
        }

        // Construtor da lista duplamente encadeada
        public DoublyLinkedList()
        {
            head = null;
            tail = null;
            _length = 0;
        }

        // Método para verificar se a lista está vazia
        public bool isEmpty() => _length == 0;

        // Método para obter o comprimento da lista
        public int length() => _length;

        // Método para exibir a lista do início ao fim
        public void displayForward()
        {
            if (head == null)
                return;

            ListNode temp = head;
            while (temp != null)
            {
                Console.Write(temp.data + " --> ");
                temp = temp.next;
            }
            Console.WriteLine("null");
        }

        // Método para exibir a lista do fim ao início
        public void displayBackward()
        {
            if (tail == null)
                return;

            ListNode temp = tail;
            while (temp != null)
            {
                Console.Write(temp.data + " --> ");
                temp = temp.previous;
            }
            Console.WriteLine("null");
        }

        // Método para inserir um nó no início da lista
        public void insertFirst(int value)
        {
            ListNode newNode = new ListNode(value);
            if (isEmpty())
            {
                tail = newNode;
            }
            else
            {
                head.previous = newNode;
            }
            newNode.next = head;
            head = newNode;
            _length++;
        }

        // Método para inserir um nó no fim da lista
        public void insertEnd(int value)
        {
            ListNode newNode = new ListNode(value);
            if (isEmpty())
            {
                head = newNode;
            }
            else
            {
                tail.next = newNode;
                newNode.previous = tail;
            }
            tail = newNode;
            _length++;
        }

        // Método para excluir o primeiro nó da lista
        public ListNode deleteFirst()
        {
            if (isEmpty())
            {
                throw new InvalidOperationException();
            }

            ListNode temp = head;
            if (head == tail)
            {
                tail = null;
            }
            else
            {
                head.next.previous = null;
            }
            head = head.next;
            temp.next = null;
            _length--;
            return temp;
        }

        // Método para excluir o último nó da lista
        public ListNode deleteLast()
        {
            if (isEmpty())
            {
                throw new InvalidOperationException();
            }

            ListNode temp = tail;
            if (head == tail)
            {
                head = null;
            }
            else
            {
                tail.previous.next = null;
            }
            tail = tail.previous;
            temp.previous = null;
            _length--;
            return temp;
        }
    }
}
